staging/vhost-device-can/src/backend.rs

```diff
@@ -0,0 +1,352 @@
+
+    if (can_devices.len() as u32) != input.socket_count {
+        info!(
+            "Number of CAN/FD devices ({}) not equal with socket count {}",
+            input.can_devices, input.socket_count
+        );
+        return Err(Error::SocketCountInvalid(
+            input.socket_count.try_into().unwrap(),
+        ));
+    }
+
+    match check_can_devices(&can_devices) {
+        Ok(_) => Ok(can_devices),
+        Err(_) => Err(Error::CouldNotFindCANDevs),
+    }
+}
+
+impl TryFrom<CanArgs> for VuCanConfig {
+    type Error = Error;
+
+    fn try_from(args: CanArgs) -> Result<Self> {
+        if args.socket_count == 0 {
+            return Err(Error::SocketCountInvalid(0));
+        }
+
+        let can_devices = match parse_can_devices(&args) {
+            Ok(can_devs) => can_devs,
+            Err(e) => return Err(e),
+        };
+
+        Ok(VuCanConfig {
+            socket_path: args.socket_path,
+            socket_count: args.socket_count,
+            can_devices,
+        })
+    }
+}
+
+impl VuCanConfig {
+    pub fn generate_socket_paths(&self) -> Vec<PathBuf> {
+        let socket_file_name = self
+            .socket_path
+            .file_name()
+            .expect("socket_path has no filename.");
+        let socket_file_parent = self
+            .socket_path
+            .parent()
+            .expect("socket_path has no parent directory.");
+
+        let make_socket_path = |i: u32| -> PathBuf {
+            let mut file_name = socket_file_name.to_os_string();
+            file_name.push(std::ffi::OsStr::new(&i.to_string()));
+            socket_file_parent.join(&file_name)
+        };
+
+        (0..self.socket_count).map(make_socket_path).collect()
+    }
+}
+
+// This is the public API through which an external program starts the
+/// vhost-device-can backend server.
+pub(crate) fn start_backend_server(socket: PathBuf, can_devs: String) -> Result<()> {
+    loop {
+        let controller =
+            CanController::new(can_devs.clone()).map_err(Error::CouldNotCreateCanController)?;
+        let lockable_controller = Arc::new(RwLock::new(controller));
+        let vu_can_backend = Arc::new(RwLock::new(
+            VhostUserCanBackend::new(lockable_controller.clone())
+                .map_err(Error::CouldNotCreateBackend)?,
+        ));
+        lockable_controller
+            .write()
+            .unwrap()
+            .open_can_socket()
+            .map_err(Error::FailCreateCanControllerSocket)?;
+
+        let read_handle = CanController::start_read_thread(lockable_controller.clone());
+
+        let mut daemon = VhostUserDaemon::new(
+            String::from("vhost-device-can-backend"),
+            vu_can_backend.clone(),
+            GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+        )
+        .map_err(Error::CouldNotCreateDaemon)?;
+
+        /* Start the read thread -- need to handle it after termination */
```

Comment:
Please use `//` for comments

Reply:
Thanks for the comments, I have changed them all to `//`